licence-checker/licence-checker.py

```diff
@@ -157,13 +161,33 @@ def full_line_parts(self, licence_line):
 
 class LicenceMatcher:
     '''An object to match a given licence at the start of a file'''
-    def __init__(self, comment_style, licence):
+    def __init__(self, comment_style, licence, match_regex):
         self.style = comment_style
+        self.expected_lines = list()
+        # In case we are using regex matching we can pass the full line "as is"
+        if match_regex:
+            try:
+                for i, ll in enumerate(licence):
+                    self.expected_lines.append(
+                        re.compile(comment_style.expected_full_line(ll)))
+                # use the "first line" as a licence marker
+                self.search_marker = self.expected_lines[0]
+            # Catch any regex error here and raise a runtime error.
+            except re.error as e:
+                raise RuntimeError(
+                    "Cannot compile line {} of the licence as a regular expression. Saw `{}`: {}"
+                    .format(i, e.pattern[e.pos], e.msg))
```

Comment:
I had to read this a couple of times to convince myself it would work! Maybe this would be easier to understand if the try/except was inside the for loop? That way, you're not dependent on Python's rather odd lexical scoping rules. (Just a matter of switching lines 169 and 170 and re-indenting)

Reply:
Agree, hopefully, that is more readable now!